libs/compact-map/src/Data/Compact/KeyMap.hs

```diff
@@ -502,218 +522,229 @@ array1 x = fst (withMutArray 1 (\marr -> mwrite marr 0 x))
+      where
+        leafF2 k v = case searchPath k (drop n (keyPath k)) x of
+          Nothing -> ans
+          Just u -> accum ans k u v
+        arrayF2 bm2 arr2 = foldl' accum2 ans (bitmapToList bm)
+          where
+            bm = bm1 .&. bm2
+            accum2 result i =
+              foldIntersect2
+                (n + 1)
+                accum
+                result
+                (index arr1 (indexFromSegment bm1 i))
+                (index arr2 (indexFromSegment bm2 i))
+
+foldOverIntersection :: (ans -> Key -> u -> v -> ans) -> ans -> KeyMap u -> KeyMap v -> ans
+foldOverIntersection accum ans x1 x2 = foldIntersect2 0 accum ans x1 x2
 
 -- =========================================================
 
--- | Domain restrict 'hkm' to those Keys found in 's'. This algorithm
+-- | Domain restrict 'hm' to those Keys found in 's'. This algorithm
 --   assumes the set 's' is small compared to 'hm'.
-domainRestrict :: KeyMap v -> Set Key -> KeyMap v
-domainRestrict hm s = Set.foldl' accum Empty s
+--   when that is not the case, intersection variants can be used.
+restrictKeys :: KeyMap v -> Set Key -> KeyMap v
+restrictKeys hm s = Set.foldl' accum Empty s
   where
     accum ans key =
       case lookupHM key hm of
         Nothing -> ans
         Just v -> insert key v ans
 
-hmdr :: KeyMap Int
-hmdr = fromList (take 10 pairs)
+withoutKeys :: KeyMap v -> Set Key -> KeyMap v
+withoutKeys hm s = Set.foldl' accum hm s
+  where
+    accum ans key =
+      case lookupHM key hm of
+        Nothing -> ans
+        Just _ -> delete key ans
 
-set :: Set Key
-set = Set.fromList [bpairs !! 3, bpairs !! 8, bpairs !! 20]
+-- ===========================================================
+-- Maximum and Minimum Key
+
+-- | Get the smallest key, NOT the smallest value
+lookupMin :: KeyMap v -> Maybe (Key, v)
+lookupMin Empty = Nothing
+lookupMin (Leaf k v) = Just (k, v)
+lookupMin (One _ x) = lookupMin x
+lookupMin (Two _ x _) = lookupMin x
+lookupMin (BitmapIndexed _ arr) = lookupMin (index arr 0)
+lookupMin (Full arr) = lookupMin (index arr 0)
+
+-- | Get the largest key, NOT the largest value
+lookupMax :: KeyMap v -> Maybe (Key, v)
+lookupMax Empty = Nothing
+lookupMax (Leaf k v) = Just (k, v)
+lookupMax (One _ x) = lookupMax x
+lookupMax (Two _ _ y) = lookupMax y
+lookupMax (BitmapIndexed _ arr) = lookupMax (index arr (isize arr - 1))
+lookupMax (Full arr) = lookupMax (index arr (isize arr - 1))
+
+-- | The view of the KeyMap of the smallestKey and its value, and the map that results from removing that Leaf.
+minViewWithKeyHelp :: KeyMap a -> (KeyMap a -> KeyMap a) -> Maybe ((Key, a), KeyMap a)
+minViewWithKeyHelp x continue = case3 Nothing leafF arrayF x
+  where
+    leafF k v = Just ((k, v), continue Empty)
+    arrayF bm arr = minViewWithKeyHelp (index arr slicepoint) (continue . largeSide i bmMinusi slicepoint arr)
+      where
+        slicepoint = 0
+        seglist = bitmapToList bm
+        i = head seglist
```

Comment:
Same thing here.

IMHO `head` should be removed from `base`, so no one could use it.

Reply:
handled as described above, there was also a 3rd case which was handed the same way.